source/transfer_list.rst

```diff
@@ -130,20 +130,15 @@ The TE header is defined in :numref:`tab_te_header`.
      - 0x0
      - The entry type identifier.
 
-   * - hdr_size
-     - 0x4
-     - 0x4
-     - The size of this entry header in bytes.
-
    * - data_size
+     - 0x3
```

Comment:
It is better to have a data size of 4 bytes, since it avoids limiting things unnecessarily. We cannot know what sort of tables might be passed between firmware with this mechanism.

To steal a byte for the header size, we could get it from the tag, perhaps.

Reply:
I think it is OK to have an 8-byte header instead of 16. The header-size field does not need to be 4 bytes, but I believe the data-size field should be

Reply:
Ok dear, I just remembered why that is not good. It means that the contents of the TE are only 8-byte-aligned instead of 16. This won't work for ACPI tables, for example. In fact it is then impossible to have 16-byte alignment for a TE.

Reply:
We can define the ACPI entry as having a 8 byte reserved field before the data.
That way we ensure that any ACPI table is always 16-byte aligned.

Eating into the tag_id and the data_size does have potentially nasty consequences:
 - tag_id: we should have enough reserved space there if we need to partition the tag space in the future.
 - data_size: 32MB seems fine today but may not age well and be a limiting technical factor a few years from now.

Reply:
Re ACPI alignment, yes adding 8 bytes of padding that would work. It is unfortunate though, since it means that the client of the TL library will need to know to adjust the pointer / pad the size. Still, it is workable.

It seems to me that part of the drive to an 8-byte header is the intention to move bl_params (and coreboot tables?) to use individual TEs. I don't believe that is a good idea.

Reply:
Maybe wait for the outcome of the tag allocation discussion before doing this, since a sparse tag id space might be problematic with 24 bits

Reply:
I don't think it should be. The way I think sparse tag allocation should work (and as I tried to describe it in this PR) was not that we should strictly reserve large swathes of the tag space in advance with some grand organization scheme that must never be changed again afterwards, rather that tag ranges can be easily grabbed ad hoc as new tags are being added. I would suggest maybe 256 tags as the size of each range to begin with. So once people start using this specification, let's say MediaTek wants to implement 3 tags for passing on chipset-specific data, and they happen to be the first, then they'd grab the tags 0x100, 0x101 and 0x102. Next maybe the coreboot project decides to port its existing coreboot table handoff interface to this spec with all the existing tags from there, so they grab 0x200 - 0x244. Then we port the TF-A bl_aux_param interface: the "generic" aux param tags get a new range starting at 0x300, the MeidaTek specific tag can go to the MediaTek range at 0x103, and the Rockchip-specific tags can get a new range at 0x400.

With 2^24 tags we have 65536 ranges like that, which should be more than more than enough to fulfill any need any stakeholder will ever get. If 256 tags aren't enough -- no big deal, just create a second range for that same stakeholder (e.g. if coreboot filled all tags from 0x200 - 0x2ff and still needs more tags, just keep going at 0x500). It doesn't need to be perfect, the goal is just to have _some_ amount of organization to make the list a little easier to read and search through afterwards.

Reply:
OK, you've convinced me that 24bits should be enough for a loose allocation scheme like this. Nobody seems to be proposing formal carveouts.

Reply:
Updated in latest push.